### PR TITLE
MF-1454 - Add Policies for sharing a Thing

### LINKED ISSUE #1454: Policies for Things sharing
**FEATURE REQUEST**

1. Is there an open issue addressing this request? If it does, please add a "+1" reaction to the
   existing issue, otherwise proceed to step 2.

2. Describe the feature you are requesting, as well as the possible use case(s) for it.
Instead of an ownership-based approach, `Things` should be shared among entities via Policies. So that, each entity that has the policy to access a `Thing` should be able to access the `Thing`. Or, if the entity has no policy to access a Thing, the entity should not access a Thing.

3. Indicate the importance of this feature to you (must-have, should-have, nice-to-have).
must-have

Some tasks for further steps:
- [ ] After deleting a `thing`, delete the policy related to that particular `thing`. It requires adding a new gRPC endpoint.

### PR DESCRIPTION
**Pull request by buraksekili**
Signed-off-by: Burak Sekili <buraksekili@gmail.com>

### What does this do?

This PR adds an HTTP endpoint and service for sharing a Thing via Policies.

### Which issue(s) does this PR fix/relate to?
Resolves #1454 

### List any changes that modify/break current functionality
This PR updates `thingRepository`. Instead of the ownership-based approach, it uses policies.
### Have you included tests for your changes?
Yes
### Did you document any new/modified functionality?
Yes
### Notes

The HTTP endpoint for sharing a thing can be tested via:

```bash
curl -isSX POST http://localhost/things/<thing_id>/share -d '{"user_ids":["<user_id>"], "policies": ["read", "write", "delete"]}' -H "Authorization: <user_auth_token>" -H 'Content-Type: application/json'
```

### CODE REVIEW

File: things/service.go [hunk 1]
```diff
@@ -247,13 +242,46 @@ func (ts *thingsService) UpdateThing(ctx context.Context, token string, thing Th
...
+				errs = accumulateError(errs, ErrAuthorization, thingID, userID)
+			}
+		}
+	}
+	return errs
+}
+
+func accumulateError(errs, newErr error, thingID, userID string) error {
+	return errors.Wrap(newErr, errors.Wrap(fmt.Errorf("cannot claim ownership on thing: %s by user: %s", thingID, userID), errs))
+}
```
Comment (dborovcanin) [inline things/service.go]: No need for a separate function and there is no need to wrap each error to `ErrAuthorization`. Something like:
```go
func (ts *thingsService) claimOwnership(ctx context.Context, thingID string, policies, userIDs []string) error {
	var errs error
	for _, userID := range userIDs {
		for _, policy := range policies {
			apr, err := ts.auth.AddPolicy(ctx, &mainflux.AddPolicyReq{Obj: thingID, Act: policy, Sub: userID})
			if err != nil {
                                errs = errors.Wrap(fmt.Errorf("Cannot claim ownership on thing '%s' by user '%s': %s", thingID, userID, err), errs))
			}
			if !apr.GetAuthorized() {
				errs = errors.Wrap(fmt.Errorf("Cannot claim ownership on thing '%s' by user '%s': unauthorized", thingID, userID), errs))
			}
		}
	}
        if errs != nil {
                return errors.Wrap(ErrAuthorization, errs)
        }
	return nil
}
```
will do.

File: things/service_test.go [hunk 2]
```diff
@@ -154,6 +155,55 @@ func TestUpdateKey(t *testing.T) {
...
 
+func TestShareThing(t *testing.T) {
+	email2 := "user2@example.com"
+	svc := newService(map[string]string{token: email, token2: email2})
+	ths, err := svc.CreateThings(context.Background(), token, thing)
+	require.Nil(t, err, fmt.Sprintf("unexpected error: %s\n", err))
+	th := ths[0]
+	policies := []string{"read"}
+
+	cases := []struct {
```
Comment (dborovcanin) [inline things/service_test.go:166]: Test for partial success (some failed, some successfully shared) is missing.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: things/service_test.go [hunk 3]
```diff
@@ -154,6 +155,63 @@ func TestUpdateKey(t *testing.T) {
 	}
 }
 
+func TestShareThing(t *testing.T) {
+	email2 := "user2@example.com"
+	svc := newService(map[string]string{token: email, token2: email2})
+	ths, err := svc.CreateThings(context.Background(), token, thing)
+	require.Nil(t, err, fmt.Sprintf("unexpected error: %s\n", err))
+	th := ths[0]
+	policies := []string{"read"}
```
Comment (dborovcanin) [inline things/service_test.go]: Change order; prepare test (create service, create things, require nil erro) and declare variables for different cases after:
```go
	svc := newService(map[string]string{token: email, token2: email2})
	ths, err := svc.CreateThings(context.Background(), token, thing)
	require.Nil(t, err, fmt.Sprintf("unexpected error: %s\n", err))
	th := ths[0]
        email2 := "user2@example.com"
	policies := []string{"read"}
```

File: things/service.go [hunk 4]
```diff
@@ -247,13 +242,42 @@ func (ts *thingsService) UpdateThing(ctx context.Context, token string, thing Th
 	return ts.things.Update(ctx, thing)
 }
 
+func (ts *thingsService) ShareThing(ctx context.Context, token, thingID string, policies, userIDs []string) error {
```
Comment (dborovcanin) [inline things/service.go]: Is `policies` parameter named ambiguously? It represents the action, not a list of `policies` per se, it's not "policy ids" or "policy names"?

Comment (buraksekili) [reply things/service.go]: Yes, it represents actions related to Things policies: `access`, `write` and `delete`. However, in the essence, they are `relations (actions)`, not policies. I'll update the naming.

**Review by dborovcanin — APPROVED** [review]

**Review by blokovi — APPROVED** [review]

**Pull request merged**